core/src/main/java/com/linecorp/armeria/common/stream/StreamMessageInputStream.java

```diff
@@ -0,0 +1,171 @@
+
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.util.Exceptions;
+import com.linecorp.armeria.internal.common.stream.StreamMessageUtil;
+
+import io.netty.util.concurrent.EventExecutor;
+
+final class StreamMessageInputStream<T> extends InputStream {
+
+    private final StreamMessage<T> source;
+    private final EventExecutor executor;
+    private final StreamMessageInputStreamSubscriber<T> subscriber;
+    private final AtomicBoolean subscribed = new AtomicBoolean();
+    private volatile boolean closed;
+
+    StreamMessageInputStream(StreamMessage<T> source,
+                             Function<? super T, ? extends HttpData> httpDataConverter,
+                             EventExecutor executor) {
+        requireNonNull(source, "source");
+        requireNonNull(httpDataConverter, "httpDataConverter");
+        requireNonNull(executor, "executor");
+        this.source = source;
+        this.executor = executor;
+        subscriber = new StreamMessageInputStreamSubscriber<>(httpDataConverter);
+    }
+
+    @Override
+    public int read() throws IOException {
+        return read(in -> {
+            try {
+                return in.read();
+            } catch (IOException e) {
+                return Exceptions.throwUnsafely(e);
+            }
+        });
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        return read(in -> {
+            try {
+                return in.read(b, off, len);
+            } catch (IOException e) {
+                return Exceptions.throwUnsafely(e);
+            }
+        });
+    }
+
+    private int read(Function<InputStream, Integer> function) throws IOException {
+        ensureOpen();
+        ensureSubscribed();
+        subscriber.request();
+        return function.apply(byteBufsInputStream());
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (closed) {
+            return;
+        }
+        closed = true;
+        source.abort();
+        byteBufsInputStream().close();
+    }
+
+    @Override
+    public int available() throws IOException {
+        ensureOpen();
+        return byteBufsInputStream().available();
+    }
+
+    private void ensureOpen() throws IOException {
+        if (closed) {
+            throw new IOException("Stream closed");
+        }
+    }
+
+    private void ensureSubscribed() {
+        if (subscribed.compareAndSet(false, true)) {
+            source.subscribe(subscriber, executor);
+        }
+        subscriber.whenSubscribed.join();
```

Comment:
Should we move `subscriber.whenSubscribed.join()` inside `if` condition?

Reply:
✅ updated.